common/src/main/java/com/wynntils/core/chat/ChatManager.java

```diff
@@ -0,0 +1,221 @@
+            // First remove the "Press SHIFT to continue" trailer.
+            newLines.removeFirst();
+            if (newLines.getFirst().isEmpty()) {
+                newLines.removeFirst();
+            } else {
+                WynntilsMod.warn("Malformed dialog [#1]: " + newLines.getFirst());
+            }
+
+            // Separate the dialog part from any potential new "real" chat lines
+            boolean dialogDone = false;
+            for (String line : newLines) {
+                if (!dialogDone) {
+                    if (line.equals("§r")) {
+                        dialogDone = true;
+                        // Intentionally throw away this line
+                    } else {
+                        dialog.push(line);
+                    }
+                } else {
+                    newChatLines.push(line);
+                }
+            }
+        } else {
+            // After a NPC dialog screen, Wynncraft sends a "clear screen" with line of ÀÀÀ...
+            // We just ignore that part
+            if (trailingLine.matches("À+")) {
+                newLines.removeFirst();
+                if (newLines.getFirst().equals("§r")) {
+                    newLines.removeFirst();
+                } else {
+                    WynntilsMod.warn("Malformed dialog [#2]: " + newLines.getFirst());
+                }
+            }
+
+            // What remains, if any, are new chat lines
+            newLines.forEach(newChatLines::push);
+        }
+
+        // Register all new chat lines
+        newChatLines.forEach(line -> handleFakeChatLine(line));
+
+        // Update the new dialog
+        handleNpcDialog(dialog);
+    }
+
+    private static void handleFakeChatLine(String codedString) {
+        // This is a normal, single line chat but coded with format codes
+        saveLastChat(codedString);
+        TextComponent message = new TextComponent(codedString);
+        Component updatedMessage = handleChatLine(message, codedString, MessageType.BACKGROUND);
+        // If the message is canceled, we do not need to cancel any packets,
+        // just don't send out the chat message
+        if (updatedMessage == null) return;
+
+        McUtils.sendMessageToClient(updatedMessage);
+    }
+
+    private static void saveLastChat(String codedString) {
+        String msg = ComponentUtils.stripFormattingCodes(codedString);
+        if (!msg.isBlank()) {
+            lastRealChat = msg;
+        }
+    }
+
+    private static RecipientType getRecipientType(Component message, MessageType messageType) {
+        String msg = ComponentUtils.getCoded(message);
+        if (messageType == MessageType.SYSTEM) {
+            // System type messages can only be shouts or "info" messages
+            // We call this MessageType.NORMAL anyway...
+            if (RecipientType.SHOUT.matchPattern(msg, MessageType.NORMAL)) {
+                return RecipientType.SHOUT;
+            }
+            return RecipientType.INFO;
```

Comment:
Qodana suggestion: extract this return (if there was not a specific reason not to do so.)

Reply:
I'm not sure what Qodana is talking about here. There is no common part? Yes, SHOUT is one member of RecipientType.values(), but seriously, how could you extract that meaningfully?

I claim Qodana is just being stupid right now. If you see a way to simplify this code, please explain to me in better terms than Qodana is doing...

Reply:
It's about the return RecipientType.INFO. That is common. Anyways this is just a nit.

Reply:
Ah, I see, it wants to extract the return INFO.

Well, you could do that. I just think this captures the semantics better. Initially, I had SYSTEM always return INFO, until I discovered that SHOUTS where (sometimes? or possibly always?) coded as SYSTEM, so I had to put in a special case for that.

So the semantics is like:
* for system, the default case is "info" unless it is a shout
* for normal, the default case is that it's a chat message. if it's not, the fallback is that it's "info"

But if we can't silence Qodana any other way, I can rearrange this.

Reply:
Ah, it is alright then. Qodana should only report problems with added code (although this might not be the case atm, perhaps it is buggy)